EX-1850 - Implement Report query and deletion

johnsonw opened the pull request:
Implement Report query and deletion in graphql

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2304)
<!-- Reviewable:end -->

johnsonw commented:
![image](https://user-images.githubusercontent.com/1100657/94964446-d5637480-04c7-11eb-9465-a65d8dad217e.png)

![image](https://user-images.githubusercontent.com/1100657/94964381-b9f86980-04c7-11eb-96d9-22c3d0373204.png)

```
[root@adm report]# touch file1.txt
[root@adm report]# ls
file1.txt
[root@adm report]# ls
[root@adm report]#
```

jgrund commented on iml-api/src/graphql.rs:
```diff
@@ -813,6 +828,20 @@ impl MutationRoot {
+        filename: String,
+    ) -> juniper::FieldResult<bool> {
+        let mut path = get_report_path();
+        path.push(filename);
```
What happens if filename is `../../*`?

johnsonw replied:
right, I'm currently thinking about that but wanted to push a first draft. Do you have any initial thoughts?

jgrund replied:
Pushed an update.

jgrund commented on iml-api/src/graphql.rs:
```diff
@@ -467,6 +469,19 @@ impl QueryRoot {
+        let paths = tokio::fs::read_dir(get_report_path()).await?;
+
+        let items: Vec<String> = paths
+            .map_ok(|x| x.file_name().into_string().expect("DirEntry to string"))
```
Seems like we should return the error instead of panicking here.

johnsonw replied:
I'm trying to think of the best way to do this. So far I came up with:
```
let items: Vec<String> = paths
            .map_ok(|x| {
                x.file_name().into_string().map_or_else(
                    |e| format!("Error getting stratagem report file name: {:?}", e),
                    |x| x,
                )
            })
            .try_collect()
            .await?;
```
However, I feel like I'm missing something..

jgrund replied:
Pushed an update.

johnsonw commented:
Docker test:

Displaying files:
![image](https://user-images.githubusercontent.com/1100657/95108307-1b5b4b00-0709-11eb-94c3-f1042e5a90c3.png)

[root@iml-api report]# ls
[root@iml-api report]# ls
expiring_fids-fs-32621d4c-9b38-4d22-8d4a-9a45d96f5667.txt

Deleting a file:
![image](https://user-images.githubusercontent.com/1100657/95108100-cf100b00-0708-11eb-81f8-a299639fdf68.png)
```
[root@iml-api report]# ls
expiring_fids-fs-52ee437c-646a-402a-9df3-e90810a77a68.txt
[root@iml-api report]# ls
[root@iml-api report]#
```

johnsonw commented:
Trying to delete a file maliciously:

![image](https://user-images.githubusercontent.com/1100657/95109159-38444e00-070a-11eb-8fde-da2aa4f8e60c.png)

mkpankov reviewed (approved):
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2304)* status: all files reviewed, 2 unresolved discussions (waiting on @johnsonw)



<!-- Sent from Reviewable.io -->

johnsonw commented:
Integration tests passed here: https://build.whamcloud.com/view/IML/job/manager-for-lustre/2489/

johnsonw commented:
Also, I've updated the query to return both the filesize and modified time. The data returned now looks like this:
```
{
  "data": {
    "stratagemReports": [
      {
        "filename": "/var/spool/iml/report/test.txt",
        "modifyTime": "2020-10-05T21:51:57.887528595+00:00",
        "size": 0
      }
    ]
  }
}
```

johnsonw commented:
The CodeQL failure is unrelated and I see it in other patches as well.

jgrund commented:
Does this now print the whole file path or just the name?

johnsonw commented:
It's printing the whole file path. I thought it might be useful to show the whole path in case they weren't aware of the report being under /var/spool/iml. However, we could always print a note above indicating where the file live and then just show the filename. What are your thoughts?

jgrund commented:
Probably just the filename. The reports can be downloaded at /report/<filename>, so it's probably not important to know the full path. In addition, if we are running in containers, that path will exist inside a container only and will be confusing to users that it's not on the host.

johnsonw commented:
Sure, makes sense. I will update it to be the filename only.

johnsonw commented:
Here is an updated screenshot. I created three files under /var/spool/iml/report. Notice that only the filename is now displayed:
test1.txt = 1Kib
test2.txt = 1Mib
test3.txt = 1Gib

![image](https://user-images.githubusercontent.com/1100657/95228911-e95fec80-07cd-11eb-9e28-4372b91de13d.png)

ip1981 commented on iml-api/src/graphql.rs:
```diff
@@ -813,6 +855,24 @@ impl MutationRoot {
+        let path = tokio::fs::canonicalize(report_base.join(filename)).await?;
+
+        if !path.starts_with(report_base) {
+            return Err(FieldError::new("Invalid path", Value::null()));
```
nit:
```suggestion
            return Err(FieldError::new("Invalid filename", Value::null()));
```

ip1981 commented on iml-api/src/graphql.rs:
```diff
@@ -467,6 +483,32 @@ impl QueryRoot {
+            .await?
+            .into_iter()
+            .map(|filename| {
+                fs::canonicalize(get_report_path().join(filename.clone()))
```
Could call `get_report_path()` juts once at the beginning?

ip1981 replied:
Is clone() needed?

ip1981 commented on iml-api/src/graphql.rs:
```diff
@@ -467,6 +483,32 @@ impl QueryRoot {
+    /// List completed Stratagem reports that currently reside on the manager node.
+    /// Note: All report names must be valid unicode.
+    async fn stratagem_reports(_context: &Context) -> juniper::FieldResult<Vec<StratagemReport>> {
+        let paths = tokio::fs::read_dir(get_report_path()).await?;
```
nit:
```suggestion
        let paths = fs::read_dir(get_report_path()).await?;
```
